numpy/core/src/umath/reduction.c

```diff
@@ -293,98 +68,86 @@ count_axes(int ndim, const npy_bool *axis_flags)
  */
-NPY_NO_EXPORT PyArrayObject *
+NPY_NO_EXPORT int
 PyArray_InitializeReduceResult(
                     PyArrayObject *result, PyArrayObject *operand,
-                    const npy_bool *axis_flags,
-                    npy_intp *out_skip_first_count, const char *funcname)
+                    const npy_bool *axis_flags, const char *funcname,
+                    int keepdims)
 {
-    npy_intp *strides, *shape, shape_orig[NPY_MAXDIMS];
+    npy_intp shape[NPY_MAXDIMS], strides[NPY_MAXDIMS];
+    npy_intp *shape_orig = PyArray_SHAPE(operand);
+    npy_intp *strides_orig = PyArray_STRIDES(operand);
     PyArrayObject *op_view = NULL;
     int idim, ndim, nreduce_axes;
 
     ndim = PyArray_NDIM(operand);
 
-    /* Default to no skipping first-visit elements in the iteration */
-    *out_skip_first_count = 0;
-
-    /* Take a view into 'operand' which we can modify. */
-    op_view = (PyArrayObject *)PyArray_View(operand, NULL, &PyArray_Type);
-    if (op_view == NULL) {
-        return NULL;
-    }
-
     /*
-     * Now copy the subarray of the first element along each reduction axis,
-     * then return a view to the rest.
+     * Copy the subarray of the first element along each reduction axis.
      *
      * Adjust the shape to only look at the first element along
-     * any of the reduction axes. We count the number of reduction axes
-     * at the same time.
+     * any of the reduction axes. If keepdims is False remove the axes
+     * entirely.
      */
-    shape = PyArray_SHAPE(op_view);
     nreduce_axes = 0;
-    if (ndim) {
-        memcpy(shape_orig, shape, ndim * sizeof(npy_intp));
-    }
     for (idim = 0; idim < ndim; ++idim) {
         if (axis_flags[idim]) {
-            if (shape[idim] == 0) {
+            if (NPY_UNLIKELY(shape_orig[idim] == 0)) {
                 PyErr_Format(PyExc_ValueError,
-                             "zero-size array to reduction operation %s "
-                             "which has no identity",
-                             funcname);
-                Py_DECREF(op_view);
-                return NULL;
+                        "zero-size array to reduction operation %s "
+                        "which has no identity", funcname);
+                return -1;
+            }
+            if (keepdims) {
+                shape[idim] = 1;
+                strides[idim] = 0;
             }
-            shape[idim] = 1;
             ++nreduce_axes;
         }
+        else {
+            if (!keepdims) {
+                shape[idim - nreduce_axes] = shape_orig[idim];
+                strides[idim - nreduce_axes] = strides_orig[idim];
+            }
+            else {
+                shape[idim] = shape_orig[idim];
+                strides[idim] = strides_orig[idim];
+            }
+        }
     }
 
-    /*
-     * Copy the elements into the result to start.
-     */
-    if (PyArray_CopyInto(result, op_view) < 0) {
-        Py_DECREF(op_view);
-        return NULL;
+    if (!keepdims) {
```

Comment:
With the above, this could become an unconditional `ndim = idim_out`. Though I think it is probably better to use a new name (just `idim_out`??)